### ISSUE TITLE
add option to automatically sync

### ISSUE THREAD

**Issue by stuart12** [android-password-store/Android-Password-Store#277]
It would be great if Android-Password-Store automatically synced the git respository.  This is for two reasons. I'm not always connected and would like to have any changes pushed automatically when I next connect. Also, I will never remember to sync when I am connected and so will find myself with stale passwords or conflicts if I edit the same password entry on two devices.

A possible specification: if the local store is dirty Android-Password-Store should sync immediately if connected otherwise at next connection. Otherwise Android-Password-Store should pull every time the app is opened (with a minimum time between pulls). It should also pull at least once a day if connected or at the next connection. It would be good if the sync at opening could be in the background and just update the display if new data arrives.

Perhaps this should be two tickets. One which pushes if dirty and another to implement background pulling.

This should not be implemented before #276 is fixed.

**Comment by pbeckmann** [comment]
Hey,

just a thought about this feature. Yes truly this would be a nice function. But currently you need a password for the sync. And i highly discourage from running a password less sync, as then somebody could overwrite your passwords....

So in a way the app has to ask for your password. Maybe the app could look when you open it, if there is a connection and suggest a sync, if the last one is older than a certain amount of time, or the local version has unsynced commits.

Greetings

**Comment by stuart12** [comment]
Well, I don't need a password for syncing as I have a ssh-key without a password. I'd rather have the ssh-key encrypted with the same password as the password for the GPG key and have them linked but that doesn't seem to be an option at the moment.

**Comment by pbeckmann** [comment]
Yes, you have that :D My humble opinion was more a general point.

**Comment by ruvido** [comment]
A bit puzzled here. I have generated an ssh-key within the app, uploaded it to github. When I sync, I always get asked for the ssh password. Is there a way to avoid this?

**Comment by zeapo** [comment]
@ruvido You get asked for a passphrase automatically because we do not keep any information about your passphrase.
IMO, an ssh-key has to have a passphrase most of the time, especially in this case. 

We could add an option to remember the ssh-key **passphrase**, by storing it encrypted using the gpg key. But we wont' encrypt the ssh-key itself using gpg.

In the future we'd like to use a gpg subkey for ssh authentication, but that's another story.

**Comment by ibrokemypie** [comment]
Why not pull upon launch, and push whenever changes are made?

**Comment by zeapo** [comment]
@ibrokemypie This is the way I'm imagining it to be. It would also have a time limite such as one pull every hour or so.

**Comment by stuart12** [comment]
In a comment in #378, I suggest allowing the user the select an password entry in the store which is the SSH passphrase.  Once this is done, whenever the GPG key is unlocked we have the SSH passphrase. So when the GPG key is unlocked we can push if there is something to push and to pull if a timeout has expired (we could have two timeouts: one for WiFi and the other for data).

**Comment by nicobo** [comment]
Hi. Is there any news about this feature ?

I'm using [qtpass](https://qtpass.org/) on an everyday basis : it automatically refreshes updates and pushes whenever a change occurs and works great for me (git server with ssh key, no password as ssh key is for headless access)

**Comment by msfjarvis** [comment]
> Hi. Is there any news about this feature ?

Not at the moment, and it's unlikely to make it into the June release regardless — we have a lot of changes ongoing and don't have the development bandwidth. I'll consider this for the July release when the time for that comes around.

**Comment by Skrilltrax** [comment]
@msfjarvis we discussed an option to sync every time the app is launched. Want to implement it that way?

**Comment by msfjarvis** [comment]
> @msfjarvis we discussed an option to sync every time the app is launched. Want to implement it that way?

I can get behind that.

**Comment by Skrilltrax** [comment]
Cool, I can take a look at it later today.

**Comment by dm17** [comment]
Do you guys do 1 git repo per password sharing group/org?
For organizations that update and then share passwords, I'm wondering what the best process is - especially considering the manual nature of git syncing. A CI/CD-like system for each password repository in git seems like overkill, for example. But every time a password is updated, then one should certainly 'git push'. A local script could also be run so a 'git pull' is run every time pass is unlocked. Thoughts?

**Comment by akselaase** [comment]
Should this be re-opened since the feature was reverted?

**Comment by msfjarvis** [comment]
> Should this be re-opened since the feature was reverted?

Yep.

**Comment by msfjarvis** [comment]
Closing as part of [issue tracker cleanup](https://github.com/android-password-store/Android-Password-Store/discussions/1836).
